### PR TITLE
feat(cli): implement interactive shell autocompletion

### LINKED ISSUE #2492: Enhance Shell Mode with Interactive Autocompletion
### Summary
This feature request proposes the implementation of interactive, readline-style autocompletion within the CLI's shell mode (`!`).

### Motivation
The current shell mode implementation lacks any form of autocompletion. This absence significantly degrades the user experience for anyone accustomed to standard shell environments (e.g., bash, zsh, fish), introducing friction and reducing workflow efficiency. Users must manually type full commands and paths, increasing the likelihood of typos and requiring them to leave the CLI to verify names. Implementing this feature would bring the shell mode closer to parity with native terminals, making it a much more powerful and viable tool for interactive use.

### Proposed Implementation
A robust implementation should provide context-aware tab-completion for:
- **Commands:** Executables available in the system's `$PATH`.
- **File & Directory Paths:** Relative and absolute paths from the current working directory.
- **(Optional) Command Arguments:** Context-sensitive suggestions for popular tools (e.g., `git` subcommands and branch names).

### Acceptance Criteria
- Pressing the `Tab` key in shell mode with a partially typed command/path should trigger a list of suggestions.
- If only one suggestion is available, it should be completed inline.
- The completion logic should correctly handle file and directory paths, including those with spaces or special characters.
- The feature should be responsive and not introduce noticeable latency to the input prompt.

### PR DESCRIPTION
**Pull request by mrpmohiburrahman**
## Summary

This PR implements interactive, readline-style tab autocompletion within the CLI's shell mode. 

## Details

- **Core Autocompletion Logic:** Implemented a new `useShellCompletion` hook to handle parsing the current input buffer and generating appropriate suggestions.
- **File & Directory Pathing:** Added robust completion for relative and absolute paths, including accurate escaping and sorting algorithms for dotfiles, paths with spaces, and quoted paths.
- **Context-Sensitive Providers:** Developed dedicated completion providers for common developer tools, specifically `git` and `npm`, which intelligently suggest relevant subcommands and arguments.
- **UI & Navigation:** Integrated the autocomplete functionality into the `InputPrompt` component, allowing users to hit `Tab` to trigger suggestions, use arrow keys to navigate the list, and press `Enter` to accept inline completions without noticeable latency.
- **Testing:** Added comprehensive unit and integration tests across `useShellCompletion.test.ts`, `gitProvider.test.ts`, `npmProvider.test.ts`, and `InputPrompt.test.tsx` to handle various input edge cases.

## Related Issues

Closes #2492

## How to Validate

1. Start the CLI in development mode (`npm run start`).
2. Enter shell mode (`!`).
3. **Path Completion:** 
   - Type `cd ` followed by the first few letters of a local directory, then press `Tab`.
   - Test autocomplete with dotfiles (e.g., `cat .gi` + `Tab`) and directories with spaces (e.g., `cd "my fold` + `Tab`).
4. **Command Arguments:** 
   - Type `git ` and press `Tab` to see contextual git subcommands.
   - Type `npm ` and press `Tab` to see npm subcommands.
5. **Interactive Selection:** Use the arrow keys to cycle through multiple suggestion results and press `Enter` to accept the highlighted option. Run the completed command to ensure it executes correctly.
6. Verify all test suites pass by running `npm run test`. 

## Pre-Merge Checklist

- [ ] Updated relevant documentation and README (if needed)
- [x] Added/updated tests (if needed)
- [ ] Noted breaking changes (if any)
- [x] Validated on required platforms/methods:
  - [x] MacOS
    - [x] npm run
    - [ ] npx
    - [ ] Docker
    - [ ] Podman
    - [ ] Seatbelt
  - [ ] Windows
    - [ ] npm run
    - [ ] npx
    - [ ] Docker
  - [ ] Linux
    - [ ] npm run
    - [ ] npx
    - [ ] Docker

### CODE REVIEW

File: packages/cli/src/ui/hooks/useShellCompletion.ts [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+ * Debounce interval (ms) for file system completions.
+ */
+const FS_COMPLETION_DEBOUNCE_MS = 50;
+
+// Backslash-quote shell metacharacters on non-Windows platforms.
+
+// On Unix, backslash-quote shell metacharacters (spaces, parens, etc.).
+// On Windows, cmd.exe doesn't use backslash-quoting and `\` is the path
+// separator, so we leave the path as-is.
+const UNIX_SHELL_SPECIAL_CHARS = /[ '"()&|;<>!#$`{}[\]*?]/g;
```
[the remark this answers is not in the document]
Comment (mrpmohiburrahman) [reply packages/cli/src/ui/hooks/useShellCompletion.ts]: Fixed it, changes on the latest commit.

File: packages/cli/src/ui/hooks/shell-completions/gitProvider.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            .split('\n')
+            .map((b) => b.trim())
+            .filter(Boolean);
+
+          return {
+            suggestions: branches
+              .filter((b) => b.startsWith(partial))
+              .map((b) => ({
+                label: b,
+                value: b,
```
[the remark this answers is not in the document]
Comment (mrpmohiburrahman) [reply packages/cli/src/ui/hooks/shell-completions/gitProvider.ts]: Fixed it, changes on the latest commit.

File: packages/cli/src/ui/hooks/shell-completions/npmProvider.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+          typeof pkg.scripts === 'object'
+            ? Object.keys(pkg.scripts)
+            : [];
+
+        return {
+          suggestions: scripts
+            .filter((s) => s.startsWith(partial))
+            .map((s) => ({
+              label: s,
+              value: s,
```
[the remark this answers is not in the document]
Comment (mrpmohiburrahman) [reply packages/cli/src/ui/hooks/shell-completions/npmProvider.ts]: Fixed it, changes on the latest commit.

File: packages/cli/src/ui/hooks/useShellCompletion.ts [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+        if (signal.aborted) return;
+
+        const queryLower = query.toLowerCase();
+        results = pathCacheRef.current
+          .filter((cmd) => cmd.toLowerCase().startsWith(queryLower))
+          .slice(0, MAX_SHELL_SUGGESTIONS)
+          .map((cmd) => ({
+            label: cmd,
+            value: cmd,
+            description: 'command',
```
[the remark this answers is not in the document]
Comment (mrpmohiburrahman) [reply packages/cli/src/ui/hooks/useShellCompletion.ts:450]: Fixed it, changes on the latest commit.

File: packages/cli/src/ui/hooks/useCommandCompletion.tsx [hunk 5]
```diff
@@ -199,9 +248,36 @@ export function useCommandCompletion({
...
+    const tokenInfo = getTokenAtCursor(currentLine, cursorCol);
+    if (!tokenInfo) {
+      return { shellTokens: [''], shellCursorIndex: 0, shellCommandToken: '' };
+    }
+    return {
+      shellTokens: tokenInfo.tokens,
+      shellCursorIndex: tokenInfo.cursorIndex,
+      shellCommandToken: tokenInfo.commandToken,
+    };
+  }, [buffer.lines, cursorRow, cursorCol, shellModeActive]);
```
[the remark this answers is not in the document]
Comment (mrpmohiburrahman) [reply packages/cli/src/ui/hooks/useCommandCompletion.tsx]: Fixed it, changes on the latest commit.

File: packages/cli/src/ui/hooks/useShellCompletion.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrpmohiburrahman) [reply packages/cli/src/ui/hooks/useShellCompletion.ts:450]: Fixed it, changes on the latest commit.

**Comment by mrpmohiburrahman** [comment]
Here is a demo of the interactive shell autocompletion in action!

https://github.com/user-attachments/assets/19cd36e9-c574-4633-85e1-253b6babd1d2

The video demonstrates:

- Path Autocompletion: Completing local directories, including correctly escaping paths with spaces and showing hidden dotfiles.

- Context-Sensitive Providers: Generating contextual sub-command suggestions for tools like git and npm.

- Interactive Navigation: Using the arrow keys to seamlessly cycle through suggestions and pressing Enter to accept the inline completion without any latency.

Let me know if you have any feedback

**Review by jacob314 — APPROVED** [review]
This is a fantastic addition to the CLI.

### Summary
The PR successfully adds a robust shell autocompletion feature to the CLI's shell mode, including support for basic pathing and custom completions for tools like `git` and `npm`. The extraction of `useShellCompletion` and specific providers ensures `useCommandCompletion` remains manageable. The tests are comprehensive, and the implementation is clean, handling nuances like paths with spaces and various quotes seamlessly.

### Findings
*   **Critical**: None. The build, lint, and tests all pass successfully locally. The execution bounds of `child_process.execFile` in `gitProvider` and file-system reads in `npmProvider` are secure and handle edge-cases and failures correctly (e.g., catching errors and gracefully returning empty suggestions).
*   **Improvements**:
    *   In `useShellCompletion.ts`, `getTokenAtCursor` accurately isolates the token when the cursor is over characters. However, if the user places the cursor inside whitespace between words (e.g., between "git" and "status"), it currently treats the cursor as being at the very end of the line (appending to the token list) instead of matching the specific positional index of the whitespace. While this works fine for the standard append-at-end workflow, fixing it to map the index properly for mid-string edits could make completions more precise in the future.
*   **Nitpicks**:
    *   A minor update to the test snapshots in `src/ui/components/InputPrompt.test.tsx.snap` was required since Vitest marked three snapshots as obsolete locally. A quick `vitest -u` resolves this.
